### ISSUE TITLE
mos.setup_gdal(spark) not working

### ISSUE THREAD

**Issue by maxi-micheli** [databrickslabs/mosaic#388]
**Describe the bug**
When I run this function, it says it is saved in the directory as per the install instructions. However, it creates the folder but it doesn't create the init_script to be run. 

**To Reproduce**
import mosaic as mos
mos.enable_mosaic(spark, dbutils)
mos.setup_gdal(spark)

**Expected behavior**
![image](https://github.com/databrickslabs/mosaic/assets/85141766/67b6111d-07f4-4de8-b622-f6f3bd234e8e)

**Screenshots**
![image](https://github.com/databrickslabs/mosaic/assets/85141766/c71ebe5b-aff0-4c11-94f7-6ae6d85f0c18)

**Additional context**
As you can see, the folder it creates is empty

**Comment by r3stl355** [comment]
Hi, please try using a previous Mosaic release (e.g `%pip install databricks-mosaic==0.3.9`) and run on DBR prior to 13 (e.g. DBR 12.2.LTS). Fixes and upgrades are on their way.

**Comment by maxi-micheli** [comment]
I understand. However, I experience the same issue when I have to do the init_script procedure to install GDAL. How could I overcome this issue? Thank you

**Comment by milos-colic** [comment]
@maxi-micheli can you try again with the latest mosaic version?
we just released v0.3.11 with fixes around setup_gdal method.

**Comment by PadenZach** [comment]
This doesnt seem to be working on latest mosaic versions either. Just attempted to install on DBR 13.0 ML and DBR 13.0 to no avail.

```py
# Databricks notebook source
# MAGIC %pip install databricks-mosaic

# COMMAND ----------

from mosaic import enable_mosaic
enable_mosaic(spark, dbutils)

# COMMAND ----------

import mosaic as mos
mos.enable_mosaic(spark, dbutils)
mos.setup_gdal(spark)

# COMMAND ----------

!ls /dbfs/FileStore/geospatial/mosaic/gdal/

# COMMAND ----------

# Restarted and configured init script.
%pip install databricks-mosaic

# COMMAND ----------

dbutils.library.restartPython()

# COMMAND ----------

import mosaic as mos
mos.enable_mosaic(spark, dbutils)
mos.enable_gdal(spark)
```

Which results in the following output:
```
GDAL not enabled. Mosaic with GDAL requires that GDAL be installed on the cluster.

Please run setup_gdal() to generate the init script for install GDAL install.

After the init script is generated, please restart the cluster with the init script to complete the setup.

Error: An error occurred while calling z:com.databricks.labs.mosaic.gdal.MosaicGDAL.enableGDAL.
: java.lang.UnsatisfiedLinkError: Can't load library: /usr/lib/libgdal.so.30
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1835)
	at java.lang.Runtime.load0(Runtime.java:783)
	at java.lang.System.load(System.java:1100)
	at com.databricks.labs.mosaic.gdal.MosaicGDAL$.loadSharedObjects(MosaicGDAL.scala:90)
	at com.databricks.labs.mosaic.gdal.MosaicGDAL$.$anonfun$enableGDAL$1(MosaicGDAL.scala:43)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.util.Try$.apply(Try.scala:213)
	at com.databricks.labs.mosaic.gdal.MosaicGDAL$.enableGDAL(MosaicGDAL.scala:41)
	at com.databricks.labs.mosaic.gdal.MosaicGDAL.enableGDAL(MosaicGDAL.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:397)
	at py4j.Gateway.invoke(Gateway.java:306)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:195)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:115)
	at java.lang.Thread.run(Thread.java:750)
```

However; the init script is configured for the cluster and appears to run correctly (no obvious errors in event logs or driver logs as far as I can tell).

Looking at the init script itself, I did find a line that made me curious:
`# please run MosaicGDAL.copySharedObjects("/dbfs/FileStore/geospatial/mosaic/gdal/") before enabling this init script`
Is this something end users need to concern themselves with?

FWIW, I _believe_ the dbfs store has the correct items in it, but I'm not _exactly_ sure what the copySharedObjects command is refering to:

```sh
>>> dbfs ls dbfs:/FileStore/geospatial/mosaic/gdal/
libgdalalljni.so
libgdalalljni.so.30
mosaic-gdal-init.sh
```

**Comment by milos-colic** [comment]
@maxi-micheli Current versions of mosaic work only on DBRs up to 12.x.
DBR 13 is using Ubuntu 22.4, while DBRs < 13 use Ubuntu 20.4
GDAL is built for 20.4 and that is why there are issues.

Could you switch to DBR 12.2LTS.

Future versions will support DBR 13+. 
Spark and Delta also have their versions updated on DBR 13, so there are a few integrations that need to be sorted out.

**Comment by PadenZach** [comment]
Thanks, that seems to get the installation working! I'm still running into issues with using the gdal reader, even while it's listed as 'enabled' -- even when while using the files from mosaic's test cases.

Should I post the information here, or in another issue?

**Comment by milos-colic** [comment]
@PadenZach here is fine. 
Please share the error messages you are getting.

Also if it is an open data source, could you share the link to it so I can try to replicate?

**Comment by PadenZach** [comment]
```py
import mosaic as mos
mos.enable_mosaic(spark, dbutils)
mos.enable_gdal(spark)
# Above returns "GDAL enabled, GDAL 3.4.3, realeased on 2022/04/22

noaa_eri_df = spark.read.format("ogr").options(
    driverName="COG",
    vsizip=False,
    asWKB=False,
    layerNumber=1,
).load(
    "s3://noaa-eri-pds/2023_Pre_Event/EC2301a_OB_N_RGB/"
)

noaa_eri_df = spark.read.format("ogr").options(
    driverName="COG",
    vsizip=False,
    asWKB=False,
    layerNumber=1,
).load(
    "s3://noaa-eri-pds/2023_Pre_Event/EC2301a_OB_N_RGB/EC2301a_OB_N_C0812830w303730n.tif"
)

# Then I tried mounting to dbfs
dbutils.fs.mount("s3a://noaa-eri-pds/2023_Pre_Event/EC2301a_OB_N_RGB/", "/mnt/zach/EC2301a_OB_N_RGB/")

```

Both result in the same error. I've also had the same issue on every dataset I attempt to load across several formats. (EG: OGR loader, geojson, geodb, shape, etc). 

```
java.lang.NullPointerException
---------------------------------------------------------------------------
Py4JJavaError                             Traceback (most recent call last)
File <command-4287091999307642>:1
----> 1 noaa_eri_df = spark.read.format("ogr").options(
      2     driverName="COG",
      3     vsizip=False,
      4     asWKB=False,
      5     layerNumber=1,
      6 ).load(
      7     "s3://noaa-eri-pds/2023_Pre_Event/EC2301a_OB_N_RGB/EC2301a_OB_N_C0812830w303730n.tif"
      8 )

File /databricks/spark/python/pyspark/instrumentation_utils.py:48, in _wrap_function.<locals>.wrapper(*args, **kwargs)
     46 start = time.perf_counter()
     47 try:
---> 48     res = func(*args, **kwargs)
     49     logger.log_success(
     50         module_name, class_name, function_name, time.perf_counter() - start, signature
     51     )
     52     return res
```

Cluster details:
DBR 12.2 LTS
i4i.xlarge single node cluster

databricks-mosaic 0.3.11 installed via cluster libraries tab.

**Comment by milos-colic** [comment]
@PadenZach Ah yes, this is due to random-read-write-access that GDAL is putting a lock on for the files in blob.
Blob storage doesn't allow these permissions hence GDAL fails to read the file and returns a null pointer.
DBFS stored data works fine with current mosaic version.

I am working on an in-memory handling of gdal/ogr files through binary columns.
The limitation will be that it will only work with files < 2GB.

Could you just verify one thing for me.
Does spark binary format work on the location you are trying to read?

GDAL is funky with the way paths are handled as well, so once this works we will need to make sure the correct path is passed. 

For now the pattern will be to copy the file from your mount to dbfs and load it from there.

**Comment by PadenZach** [comment]
Thanks for the response, I have a few points / questions :)

### 1. Binary mode
```py
>>> noaa_eri_df_binary = spark.read.format("binaryFile").load(
>>>     "s3://noaa-eri-pds/2023_Pre_Event/EC2301a_OB_N_RGB/EC2301a_OB_N_C0812830w303730n.tif"
>>> )
>>> noaa_eri_df_binary.columns
['path', 'modificationTime', 'length', 'content']
```

Seems to work just fine in binary mode.

### 2. User Story / Feature request

One of the things that we'd like to do is use mosaic to read very large geodatabases or datasets composed of large raster files while effectively using the resources of the entire cluster. Currently we're running some of these through ogr2ogr to convert the file formats before read but this only runs on the driver and has increased I/O time.

Would this allow us to extra multiple files in parallel of a geodatabase that consists of multiple files, or let's say larger raster files where each band could be fit within some lower limit (eg: the 2gb limit you mentioned) but the whole file may not?


### 3. Loading from DBFS and Driver do not work

Loading data from DBFS and Local file systems do not work (even when copied out of mount!)

#### Case 1 - copy from mount to local drive.
```py
!cp /dbfs/mnt/zach/EC2301a_OB_N_RGB/EC2301a_OB_N_C0812830w303730n.tif ./geo/EC2301a_OB_N_C0812830w303730n.tif

noaa_eri_df = spark.read.format("ogr").options(
    driverName="COG",
    # vsizip=False,
    # asWKB=False,
    layerNumber=1,
).load(
    "file:///databricks/driver/geo/EC2301a_OB_N_C0812830w303730n.tif"
)

```

#### Case 2 Copy from DBFS mount to dbfs:
```py
!cp /dbfs/mnt/zach/EC2301a_OB_N_RGB/EC2301a_OB_N_C0812830w303730n.tif /dbfs/FileStore/EC2301a_OB_N_C0812830w303730n.tif
noaa_eri_df = spark.read.format("ogr").options(
    driverName="COG",
).load(
    "/FileStore/EC2301a_OB_N_C0812830w303730n.tif"
)
```
Error:
```
Unexpected exception formatting exception. Falling back to standard exception
Traceback (most recent call last):
  File "/databricks/python/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3378, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<command-4287091999307661>", line 4, in <module>
    noaa_eri_df = spark.read.format("ogr").options(
  File "/databricks/spark/python/pyspark/instrumentation_utils.py", line 48, in wrapper
    res = func(*args, **kwargs)
  File "/databricks/spark/python/pyspark/sql/readwriter.py", line 302, in load
    return self._df(self._jreader.load(path))
  File "/databricks/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/java_gateway.py", line 1321, in __call__
    return_value = get_return_value(
  File "/databricks/spark/python/pyspark/errors/exceptions.py", line 228, in deco
    return f(*a, **kw)
  File "/databricks/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/protocol.py", line 326, in get_return_value
    raise Py4JJavaError(
py4j.protocol.Py4JJavaError: An error occurred while calling o721.load.
: java.lang.NullPointerException
	at com.databricks.labs.mosaic.datasource.OGRFileFormat$.getDataSource(OGRFileFormat.scala:372)
	at com.databricks.labs.mosaic.datasource.OGRFileFormat$.inferSchemaImpl(OGRFileFormat.scala:403)
	at com.databricks.labs.mosaic.datasource.OGRFileFormat.inferSchema(OGRFileFormat.scala:39)
	at org.apache.spark.sql.execution.datasources.DataSource.$anonfun$getOrInferFileFormatSchema$11(DataSource.scala:232)
	at scala.Option.orElse(Option.scala:447)
	at org.apache.spark.sql.execution.datasources.DataSource.getOrInferFileFormatSchema(DataSource.scala:225)
	at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:457)
	at org.apache.spark.sql.DataFrameReader.loadV1Source(DataFrameReader.scala:378)
	at org.apache.spark.sql.DataFrameReader.$anonfun$load$2(DataFrameReader.scala:334)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:334)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:240)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:380)
	at py4j.Gateway.invoke(Gateway.java:306)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:195)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:115)
	at java.lang.Thread.run(Thread.java:750)


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/databricks/python/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 1997, in showtraceback
    stb = self.InteractiveTB.structured_traceback(
  File "/databricks/python/lib/python3.9/site-packages/IPython/core/ultratb.py", line 1112, in structured_traceback
    return FormattedTB.structured_traceback(
  File "/databricks/python/lib/python3.9/site-packages/IPython/core/ultratb.py", line 1006, in structured_traceback
    return VerboseTB.structured_traceback(
  File "/databricks/python/lib/python3.9/site-packages/IPython/core/ultratb.py", line 859, in structured_traceback
    formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,
  File "/databricks/python/lib/python3.9/site-packages/IPython/core/ultratb.py", line 812, in format_exception_as_a_whole
    frames.append(self.format_record(r))
  File "/databricks/python/lib/python3.9/site-packages/IPython/core/ultratb.py", line 730, in format_record
    result += ''.join(_format_traceback_lines(frame_info.lines, Colors, self.has_colors, lvals))
  File "/databricks/python/lib/python3.9/site-packages/stack_data/utils.py", line 145, in cached_property_wrapper
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "/databricks/python/lib/python3.9/site-packages/stack_data/core.py", line 698, in lines
    pieces = self.included_pieces
  File "/databricks/python/lib/python3.9/site-packages/stack_data/utils.py", line 145, in cached_property_wrapper
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "/databricks/python/lib/python3.9/site-packages/stack_data/core.py", line 649, in included_pieces
    pos = scope_pieces.index(self.executing_piece)
  File "/databricks/python/lib/python3.9/site-packages/stack_data/utils.py", line 145, in cached_property_wrapper
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "/databricks/python/lib/python3.9/site-packages/stack_data/core.py", line 628, in executing_piece
    return only(
  File "/databricks/python/lib/python3.9/site-packages/executing/executing.py", line 164, in only
    raise NotOneValueFound('Expected one value, found 0')
executing.executing.NotOneValueFound: Expected one value, found 0
```

This seems to be some issue with gdal not being able to read the file/determine the schema? I've tried both `COG` and `GTiff` for driverName to no avail in mosaic. Running `gdalinfo` is able to provide details for files mounted using dbutils, or files copied directly to disk. I've also ran a `chmod 777` on the local copy of the file, in case there were some obscure permission issue.

**Comment by milos-colic** [comment]
@PadenZach 

Thank you for all the extra details this is really helpful! And I think this will help me design an approach that fixes the current issues both for files < 2GB but potentially for files > 2GB (which arent spark friendly in general due to runtime buffer size limits).

Would you be interested in doing a call where we could discuss the vision for some of these capabilities and what your team's needs are?

1. Binary mode works fine:
- This confirms it is GDAL file locks that are causing issues. There are 2 approaches that solve for this.
- The approach that works for files < 2GB is coming in v0.4.x, I have a working prototype
- The approach for files > 2GB should come right after (1-2 months from now)

2. very large geodatabases 
- multi_read reader is designed for this, but due to buffer issues cant go beyond 2GB before we implement the runtime file partitioning, that would ensure files are always < 2GB at runtime but can be > 2GB on disk
- Depends on the development described in point 1

3. Loading from DBFS and Driver do not work
- I just realised you are using "ogr" reader to read "TIF" or "COG"
- There are few issues here, and I will update docs to reflect this since now that I read your code I realise this isnt obvious at all. OGR reader only reads vector formats. GDAL reader reads rasters (a subset of GDAL supported formats, but in next version we are opening it fully)
- Another issue is the path. "file:///" prefix never works for GDAL bindings we are using, we are stripping this under the hood. Also direct path to "FileStore" wont work since we are appending "dbfs:/" automatically.
- Local drive won't work since spark data sources dont operate with the local drive.
- Could you try the code below:
`!cp /dbfs/mnt/zach/EC2301a_OB_N_RGB/EC2301a_OB_N_C0812830w303730n.tif /dbfs/FileStore/EC2301a_OB_N_C0812830w303730n.tif
noaa_eri_df = spark.read.format("gdal").options(
    driverName="COG",
).load(
    "dbfs:/FileStore/EC2301a_OB_N_C0812830w303730n.tif"
)`

**Comment by milos-colic** [comment]
@PadenZach great to hear that worked.
As for the docs, thanks for pointing that out, I am redoing the whole documentation as we speak, will revise the paths and examples and fix everything up.

**Comment by milos-colic** [comment]
I will close this issue now since everything discusses seemed to be addressed.
